pkg/proxy/ipvs/proxier.go

```diff
@@ -216,7 +218,7 @@ type Proxier struct {
 	ipvsScheduler string
 	// The following buffers are used to reuse memory and avoid allocations
 	// that are significantly impacting performance.
-	iptablesData     *bytes.Buffer
+	iptablesData     *iptablesData
```

Comment:
Can we keep this as `*bytes.Buffer` only?

Logging with `iptablesData` struct the rules result in a single line array of all the rules which does not look debug-friendly. IMO multi-line output is more debuggable and If I'm running with higher verbosity I would expect that from logs.

I captured some logs for IPVS proxier and pasted them here, the second log line outputs a multi-line output for `filterRules` and `natRules` which looks better. 



```
I0409 10:37:50.068128       1 proxier.go:1460] "Restoring iptables" ipFamily="IPv6" rules=["*nat",":KUBE-SERVICES - [0:0]",":KUBE-POSTROUTING - [0:0]",":KUBE-NODE-PORT - [0:0]",":KUBE-LOAD-BALANCER - [0:0]",":KUBE-MARK-MASQ - [0:0]","-A KUBE-SERVICES -s ::1/128 -j RETURN","-A KUBE-SERVICES -m addrtype --dst-type LOCAL -j KUBE-NODE-PORT","-A KUBE-LOAD-BALANCER -j KUBE-MARK-MASQ","-A KUBE-POSTROUTING -m mark ! --mark 0x00004000/0x00004000 -j RETURN","-A KUBE-POSTROUTING -j MARK --xor-mark 0x00004000","-A KUBE-POSTROUTING -m comment --comment \"kubernetes service traffic requiring SNAT\" -j MASQUERADE --random-fully","-A KUBE-MARK-MASQ -j MARK --or-mark 0x00004000","COMMIT","*filter",":KUBE-FORWARD - [0:0]",":KUBE-NODE-PORT - [0:0]",":KUBE-PROXY-FIREWALL - [0:0]",":KUBE-SOURCE-RANGES-FIREWALL - [0:0]",":KUBE-IPVS-FILTER - [0:0]",":KUBE-IPVS-OUT-FILTER - [0:0]","-A KUBE-SOURCE-RANGES-FIREWALL -j DROP","-A KUBE-FORWARD -m comment --comment \"kubernetes forwarding rules\" -m mark --mark 0x00004000/0x00004000 -j ACCEPT","-A KUBE-FORWARD -m comment --comment \"kubernetes forwarding conntrack rule\" -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT","-A KUBE-NODE-PORT -m comment --comment \"Kubernetes health check node port\" -m set --match-set KUBE-6-HEALTH-CHECK-NODE-PORT dst -j ACCEPT","-A KUBE-IPVS-FILTER -m set --match-set KUBE-6-LOAD-BALANCER dst,dst -j RETURN","-A KUBE-IPVS-FILTER -m set --match-set KUBE-6-CLUSTER-IP dst,dst -j RETURN","-A KUBE-IPVS-FILTER -m set --match-set KUBE-6-EXTERNAL-IP dst,dst -j RETURN","-A KUBE-IPVS-FILTER -m set --match-set KUBE-6-EXTERNAL-IP-LOCAL dst,dst -j RETURN","-A KUBE-IPVS-FILTER -m set --match-set KUBE-6-HEALTH-CHECK-NODE-PORT dst -j RETURN","-A KUBE-IPVS-FILTER -m conntrack --ctstate NEW -m set --match-set KUBE-6-IPVS-IPS dst -j REJECT","COMMIT",""]
I0409 10:37:50.068199       1 proxier.go:1461] "Restoring iptables" ipFamily="IPv6" natChains=<
	*nat
	:KUBE-SERVICES - [0:0]
	:KUBE-POSTROUTING - [0:0]
	:KUBE-NODE-PORT - [0:0]
	:KUBE-LOAD-BALANCER - [0:0]
	:KUBE-MARK-MASQ - [0:0]
 > natRules=<
	-A KUBE-SERVICES -s ::1/128 -j RETURN
	-A KUBE-SERVICES -m addrtype --dst-type LOCAL -j KUBE-NODE-PORT
	-A KUBE-LOAD-BALANCER -j KUBE-MARK-MASQ
	-A KUBE-POSTROUTING -m mark ! --mark 0x00004000/0x00004000 -j RETURN
	-A KUBE-POSTROUTING -j MARK --xor-mark 0x00004000
	-A KUBE-POSTROUTING -m comment --comment "kubernetes service traffic requiring SNAT" -j MASQUERADE --random-fully
	-A KUBE-MARK-MASQ -j MARK --or-mark 0x00004000
	COMMIT
 > filterChains=<
	*filter
	:KUBE-FORWARD - [0:0]
	:KUBE-NODE-PORT - [0:0]
	:KUBE-PROXY-FIREWALL - [0:0]
	:KUBE-SOURCE-RANGES-FIREWALL - [0:0]
	:KUBE-IPVS-FILTER - [0:0]
	:KUBE-IPVS-OUT-FILTER - [0:0]
 > filterRules=<
	-A KUBE-SOURCE-RANGES-FIREWALL -j DROP
	-A KUBE-FORWARD -m comment --comment "kubernetes forwarding rules" -m mark --mark 0x00004000/0x00004000 -j ACCEPT
	-A KUBE-FORWARD -m comment --comment "kubernetes forwarding conntrack rule" -m conntrack --ctstate RELATED,ESTABLISHED -j ACCEPT
	-A KUBE-NODE-PORT -m comment --comment "Kubernetes health check node port" -m set --match-set KUBE-6-HEALTH-CHECK-NODE-PORT dst -j ACCEPT
	-A KUBE-IPVS-FILTER -m set --match-set KUBE-6-LOAD-BALANCER dst,dst -j RETURN
	-A KUBE-IPVS-FILTER -m set --match-set KUBE-6-CLUSTER-IP dst,dst -j RETURN
	-A KUBE-IPVS-FILTER -m set --match-set KUBE-6-EXTERNAL-IP dst,dst -j RETURN
	-A KUBE-IPVS-FILTER -m set --match-set KUBE-6-EXTERNAL-IP-LOCAL dst,dst -j RETURN
	-A KUBE-IPVS-FILTER -m set --match-set KUBE-6-HEALTH-CHECK-NODE-PORT dst -j RETURN
	-A KUBE-IPVS-FILTER -m conntrack --ctstate NEW -m set --match-set KUBE-6-IPVS-IPS dst -j REJECT
	COMMIT
 >
```

Reply:
I think keep the iptablesData as *bytes.Buffer is also ok, since the log after that we already output the `natChains` , `natRules `, `filterChains `, `filterRules ` which are subpart of iptablesData, WDYT? @uablrek @danwinship

Reply:
We can also remove the first log line in that case, wdyt?

Reply:
Yeah, keep the buffer. I didn't change the type when I tested it in https://github.com/kubernetes/kubernetes/pull/122979#discussion_r1503127146, (I think).

But I don't think there should be any multi-line log entries. I suspect that utilities as `grep` is used to analyze logs. Especially not in text mode. In json mode, I don't think it matters as much but should be avoided anyway. This can be controlled in the implementation of logr.Marshaler and fmt.Stringer interfaces. I haven't seen any rules against it though.

Reply:
BTW, I use [jq](https://jqlang.github.io/jq/) to analyze json logs, so multi-line log entries doesn't matter, and I wouldn't see them.

Reply:
@aroradaman I missed your comment https://github.com/kubernetes/kubernetes/pull/122979#discussion_r1557626429

I guess multi-line log entries are OK.

Reply:
Thanks! I will revert the change of `iptablesData` and remove the first log line.